coprocessor/fhevm-engine/transaction-sender/src/overprovision_gas_limit.rs

```diff
@@ -0,0 +1,52 @@
+use alloy::network::{Ethereum, TransactionBuilder};
+use alloy_provider::Provider;
+use alloy_rpc_types::TransactionRequest;
+use tracing::{debug, warn};
+
+// If `txn_request.gas` is set, overprovision it by the given percent.
+// If `txn_request.gas` is not set, estimate the gas limit and then overprovision it by the given percent.
+// If the percent is less than 100, it will be set to 100%.
+// If the gas estimation fails, it will not set the gas limit for overprovisioning and will log a warning.
+pub async fn try_overprovision_gas_limit<T: Provider<Ethereum>>(
+    txn_request: impl Into<TransactionRequest>,
+    provider: &T,
+    percent: u32,
+) -> TransactionRequest {
+    if percent < 100 {
+        warn!(
+            gas_limit_overprovision_percent = percent,
+            "Overprovisioning gas limit percent is less than 100%, setting to 100%"
+        );
+    }
+    let percent = percent.max(100);
+
+    let overprovision = |gas: u64| (gas as u128 * percent as u128 / 100) as u64;
+
+    let mut txn: TransactionRequest = txn_request.into();
+
+    let new_gas = async {
+        txn.gas
+            .or({
+                match provider.estimate_gas(txn.clone()).await {
```

Comment:
or is eager !? (not sure with all these async block :) )
Maybe you want or_else ?

Reply:
Wow, good catch! Let me redo that!